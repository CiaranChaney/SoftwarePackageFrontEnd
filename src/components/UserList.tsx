import { useEffect, useState } from 'react';
import axios from 'axios';
import { Link } from 'react-router-dom';

interface User {
    userId: number;
    username: string;
    email: string;
    role: string;
}

const UserList = () => {
    const [users, setUsers] = useState<User[]>([]);

    useEffect(() => {
        const fetchUsers = async () => {
            try {
                const token = localStorage.getItem('token');
                if (!token) {
                    console.log('No token found');
                    return;
                }

                const config = {
                    headers: { Authorization: `Bearer ${token}` }
                };

                const response = await axios.get<User[]>('http://localhost:7050/api/users', config);
                setUsers(response.data);
            } catch (error) {
                console.error('Error fetching data: ', error);
            }
        };

        fetchUsers();
    }, []);

    const handleDelete = async (id: number) => {
        try {
            const token = localStorage.getItem('token');
            if (!token) {
                console.log('No token found');
                return;
            }

            const confirmDelete = window.confirm(`Are you sure you want to delete user ${id}?`);
            if (!confirmDelete) {
                return;
            }

            const config = {
                headers: { Authorization: `Bearer ${token}` }
            };

            await axios.delete(`http://localhost:7050/api/users/${id}`, config);
            setUsers(prevUsers => prevUsers.filter(user => user.userId !== id)); // Remove deleted user from the list
        } catch (error) {
            console.error('Error deleting user: ', error);
        }
    };

    return (
        <div>
            <h2>User List</h2>
            <table className="table table-dark table-striped">
                <thead>
                <tr>
                    <th>ID</th>
                    <th>Username</th>
                    <th>Email</th>
                    <th>Role</th>
                    <th>Actions</th>
                </tr>
                </thead>
                <tbody>
                {users.map(user => (
                    <tr key={user.userId}>
                        <td>{user.userId}</td>
                        <td>{user.username}</td>
                        <td>{user.email}</td>
                        <td>{user.role}</td>
                        <td>
                            <Link className="btn btn-secondary mr-2" to={`/user/edit/${user.userId}`}>Edit</Link>
                            <button className="btn btn-danger" onClick={() => handleDelete(user.userId)}>Delete</button>
                        </td>
                    </tr>
                ))}
                </tbody>
            </table>
        </div>
    );
};

export default UserList;
