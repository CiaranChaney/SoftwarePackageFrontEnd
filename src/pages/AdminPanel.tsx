import { Route, Link, Routes } from 'react-router-dom';
import UserList from '../components/UserList';
import NavBar from "../components/NavBar";
import Favicon from "../components/Favicon";

const AdminPanel = () => {
    return (
        <><NavBar></NavBar>
            <Favicon></Favicon>
            <div className="container bg-dark text-white">
                <h1 className="pt-4 pb-2">Admin Dashboard</h1>
                <nav>
                    <ul className="nav">
                        <li className="nav-item">
                            <Link className="nav-link text-light" to="/addUser">Add User</Link>
                        </li>
                    </ul>
                </nav>
                <Routes>
                    <Route path="/" element={<UserList/>}/>

                </Routes>
            </div>
        </>
    );
};

export default AdminPanel;
