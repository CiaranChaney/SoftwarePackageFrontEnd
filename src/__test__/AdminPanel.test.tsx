import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import AdminPanel from "../pages/AdminPanel";
import UserList from "../components/UserList";
import '@testing-library/jest-dom';
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import {act} from "react-dom/test-utils";

// Create a mock instance for axios
const mock = new MockAdapter(axios);

beforeEach(() => {
    // Mock NavBar and UserList if needed
    jest.mock("../components/NavBar", () => () => <nav>NavBar</nav>);
    jest.mock("../components/UserList", () => () => <div>Footer</div>);

    const mockToken = 'eyJhbGciOiJIUzI1NiJ9.eyJSb2xlIjoiQURNSU4iLCJJc3N1ZXIiOiJJc3N1ZXIiLCJVc2VybmFtZSI6IlRlc3QiLCJleHAiOjE3MTQ3Mzc5MTksImlhdCI6MTcxNDczNzkxOX0.fMN-b_fdtuHZmzMxikZzobdqUY7_YVCIEh5-qiOSp1s\n';
    Storage.prototype.getItem = jest.fn(() => mockToken); // Mock getItem method of localStorage

    // Setup Axios mocks
    mock.onGet("https://ciaranchaney.com:443/api/users").reply(200, [
        {
            userId: 1,
            username: 'testUser',
            email: 'test@example.com',
            role: 'user'
        }
    ]);
});

afterEach(() => {
    mock.resetHandlers();
    jest.restoreAllMocks();
});

describe('AdminPage', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <BrowserRouter>
                <AdminPanel />
                <UserList />
            </BrowserRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('includes necessary subcomponents', async () => {
        await act(async () => {
            render(
                <BrowserRouter>
                    <AdminPanel />
                    <UserList />
                </BrowserRouter>
            );
        });

        expect(screen.getByText('Library Hash Repository')).toBeInTheDocument();
        expect(screen.getByText('Admin Dashboard')).toBeInTheDocument();
        expect(screen.getByText('Add User')).toBeInTheDocument();
        await waitFor(() => {
            const idFields = screen.getAllByText('ID');
            idFields.forEach(field => {
                expect(field).toBeInTheDocument();
            });

            const emailFields = screen.getAllByText('Email');
            emailFields.forEach(field => {
                expect(field).toBeInTheDocument();
            });

            const roleFields = screen.getAllByText('Role');
            roleFields.forEach(field => {
                expect(field).toBeInTheDocument();
            });

            const actionFields = screen.getAllByText('Actions');
            actionFields.forEach(field => {
                expect(field).toBeInTheDocument();
            });
        });
    });

    it('renders UserList and checks for elements', async () => {
        await act(async () => {
            render(
                <BrowserRouter>
                    <AdminPanel />
                    <UserList />
                </BrowserRouter>
            );
        });


        await waitFor(() => {

            const addUserButtons = screen.getAllByText('Add User');
            addUserButtons.forEach(field => {
                expect(field).toBeInTheDocument();
            });

            const editButtons = screen.getAllByText('Edit');
            editButtons.forEach(field => {
                expect(field).toBeInTheDocument();
            });

            const deleteButtons = screen.getAllByText('Delete');
            deleteButtons.forEach(field => {
                expect(field).toBeInTheDocument();
            });

        });

    });
});
