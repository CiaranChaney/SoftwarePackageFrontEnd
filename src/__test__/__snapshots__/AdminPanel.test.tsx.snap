// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="navCol col-lg-6 mb-3"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <a
              class="link-light"
              href="/"
            >
              <h4>
                Library Hash Repository
              </h4>
            </a>
          </div>
        </div>
      </div>
      <div
        class="navCol col-lg-6 d-flex justify-content-end mb-3"
      >
        <div
          class="row"
        >
          <div
            class="col-9 col-lg-4 col-sm-6 d-flex justify-content-end mb-2 mb-lg-0"
          >
            <a
              class="btn btn-light btn-lg p-2"
              href="/library"
              role="button"
            >
              Libraries
            </a>
          </div>
          <div
            class="col-3 col-lg-3 col-sm-6 d-flex justify-content-end mb-2 mb-lg-0"
          >
            <a
              class="btn btn-light btn-lg"
              role="button"
            >
              Logout
            </a>
          </div>
          <div
            class="col-12 col-lg-5 d-flex justify-content-end"
          >
            <div
              class="text-end"
            >
              <h6>
                Logged in as: 
                <strong />
              </h6>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="container bg-dark text-white"
  >
    <h1
      class="pt-4 pb-2"
    >
      Admin Dashboard
    </h1>
    <nav>
      <ul
        class="nav"
      >
        <li
          class="nav-item"
        >
          <a
            class="nav-link text-light"
            href="/AdminPanel/user/new"
          >
            Add User
          </a>
        </li>
      </ul>
    </nav>
    <div>
      <h2>
        User List
      </h2>
      <table
        class="table table-dark table-striped"
      >
        <thead>
          <tr>
            <th>
              ID
            </th>
            <th>
              Username
            </th>
            <th>
              Email
            </th>
            <th>
              Role
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
  <div>
    <h2>
      User List
    </h2>
    <table
      class="table table-dark table-striped"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            Username
          </th>
          <th>
            Email
          </th>
          <th>
            Role
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</DocumentFragment>
`;
